core/src/cluster_info.rs

```diff
@@ -1335,55 +1359,137 @@ impl ClusterInfo {
-                            // addrs.
-                            packets
-                                .packets
-                                .push(Packet::from_data(&from_addr, protocol))
+            .filter_map(|(responses, from_addr)| {
+                if !from_addr.ip().is_unspecified()
+                    && from_addr.port() != 0
+                    && !responses.is_empty()
+                {
+                    Some((responses, from_addr))
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        if pull_responses.is_empty() {
+            return None;
+        }
+
+        let mut stats: Vec<_> = pull_responses
+            .iter()
+            .enumerate()
+            .map(|(i, (responses, _from_addr))| {
+                if let Some(stake) = stakes.get(&responses[0].pubkey()) {
+                    responses
+                        .iter()
+                        .enumerate()
+                        .map(|(j, _response)| ResponseScore {
+                            x: i,
+                            y: j,
+                            score: *stake,
                         })
+                        .collect()
                 } else {
-                    trace!("Dropping Gossip pull response, as destination is unknown");
+                    vec![ResponseScore {
+                        x: i,
+                        y: 0,
+                        score: 1,
+                    }]
                 }
-            });
+            })
+            .flatten()
+            .collect();
+
+        stats.sort_by(|a, b| a.score.cmp(&b.score));
+        let weights: Vec<_> = stats.iter().map(|stat| stat.score).collect();
+
+        let seed = [48u8; 32];
+        let rng = &mut ChaChaRng::from_seed(seed);
+        let weighted_index = WeightedIndex::new(weights).unwrap();
+
+        let mut packets = Packets::new_with_recycler(recycler.clone(), 64, "handle_pull_requests");
+        let mut total_bytes = 0;
+        let outbound_budget = me.read().unwrap().outbound_budget.bytes;
+        let mut sent = HashSet::new();
+        while sent.len() < stats.len() {
+            let index = weighted_index.sample(rng);
+            if sent.contains(&index) {
+                continue;
+            }
+            sent.insert(index);
+            let stat = &stats[index];
+            let from_addr = pull_responses[stat.x].1;
+            let response = pull_responses[stat.x].0[stat.y].clone();
+            let protocol = Protocol::PullResponse(self_id, vec![response]);
+            packets
+                .packets
+                .push(Packet::from_data(&from_addr, protocol));
+            let len = packets.packets.len();
+            total_bytes += packets.packets[len - 1].meta.size;
+
+            if total_bytes > outbound_budget {
+                break;
```

Comment:
This is where we run out of budget right?  Does this deserve a counter?

Reply:
Yep, added.